Add `BaseBodycentricRepresentation`

cmarmo opened the pull request:
### Description
This pull request adds the possibility to create geodetic representations described by [planetocentric latitudes]
~~(https://en.wikipedia.org/wiki/Latitude#Geodetic_and_geocentric_latitudes), west positive longitudes, longitudes spanning from 0 to 360 degrees.~~
The default behavior is backward compatible.
Suggested in https://github.com/astropy/astropy/issues/11170#issuecomment-1543528585.

Towards #11170.

This work is funded by the Europlanet 2024 Research Infrastructure (RI) Grant.

cmarmo commented:
It seems like I'm still confused about units conversions... :(

cmarmo commented:
For the sake of completeness, this pull request makes the `BaseGeodeticRepresentation` closer to the Planetary Data System class [Geodetic_Model](https://pds.nasa.gov/datastandards/documents/dd/current/PDS4_PDS_DD_1J00/webhelp/all/#ch06s10.html).

mhvk commented on astropy/coordinates/representation/geodetic.py:
```diff
@@ -10,6 +11,16 @@
+    elif positive_longitude == "west":
+        orient = -1
+    return Longitude(
+        orient * longitude, u.deg, wrap_angle=wrap_angle * u.deg, copy=False
```
This way you always make a copy of the longitude, which costs extra time and memory (see main comments).

mhvk replied:
Also, `wrap_angle` should be a `Quantity` already!

mhvk commented on astropy/coordinates/representation/geodetic.py:
```diff
@@ -45,9 +56,16 @@ class BaseGeodeticRepresentation(BaseRepresentation):
     to quantities holding correct values (with units of length and dimensionless,
     respectively), or alternatively an ``_ellipsoid`` attribute to the relevant ERFA
     index (as passed in to `erfa.eform`).
+    Longitudes are east positive and span from -180 to 180 degrees by default.
```
I think we should have separate classes; see main comment.

mhvk commented on astropy/coordinates/representation/geodetic.py:
```diff
@@ -45,9 +56,16 @@ class BaseGeodeticRepresentation(BaseRepresentation):
 
     attr_classes = {"lon": Longitude, "lat": Latitude, "height": u.Quantity}
+    _positive_longitude = "east"
+    _wrap_angle = 180
```
Do add a unit here.

mhvk commented on astropy/coordinates/representation/geodetic.py:
```diff
@@ -62,6 +80,16 @@ def __init_subclass__(cls, **kwargs):
+                f"Invalid argument '{cls._positive_longitude}' for '_positive_logitude' "
+                "attribute: valid argument are 'east' or 'west'."
+            )
+        if cls._wrap_angle not in [180, 360]:
```
I don't think this test is necessary; if people want to have a system with (-90, 270), why should we stop them?

mhvk commented on astropy/coordinates/tests/test_geodetic_representations.py:
```diff
@@ -29,8 +29,45 @@ class CustomGeodetic(BaseGeodeticRepresentation):
+    _flattening = 0.5886007555512007 * u.percent
+
+
+class IAUMARS2000GeodeticRepresentationWest180(BaseGeodeticRepresentation):
```
So, with my suggestion of a mixin class, this would be
`class ...(WestLongitude, BaseGeodecticRepresentation):`

mhvk commented on astropy/coordinates/tests/test_geodetic_representations.py:
```diff
@@ -29,8 +29,45 @@ class CustomGeodetic(BaseGeodeticRepresentation):
+    _wrap_angle = 360
+
+
+class IAUMARS2000GeodeticRepresentationOcentric(BaseGeodeticRepresentation):
```
And this would be `class ...(BasePlanetoCentricRepresentation)`.

mhvk commented on astropy/coordinates/tests/test_geodetic_representations.py:
```diff
@@ -29,8 +29,45 @@ class CustomGeodetic(BaseGeodeticRepresentation):
+    _equatorial_radius = 3396190.0 * u.m
+    _flattening = 0.5886007555512007 * u.percent
+    _ographic = False
+
```
Should one also test planetocentric with west longitude?

mhvk commented on astropy/coordinates/tests/test_geodetic_representations.py:
```diff
@@ -159,4 +203,103 @@ def test_geodetic_subclass_missing_equatorial_radius():
+        (
+            IAUMARS2000GeodeticRepresentationEast180.from_representation(
+                initial_cartesian
+            ).lon.value
```
Just get `.lon` - there should be no need to get `.value` and then multiple with `u.deg`.

Also, I think here and below I would just check the output directly (with numbers) rather than rely on math.

mhvk commented on astropy/coordinates/tests/test_representation.py:
```diff
@@ -1866,11 +1866,12 @@ def test_represent_as(self):
 
         # make sure represent_as() passes through the differentials
         for name in REPRESENTATION_CLASSES:
+            print(name)
```
Don't forget to remove this!

cmarmo replied:
:facepalm:

mhvk reviewed (changes requested):
@cmarmo - Good to have these options, but I do think at least some of them should be subclasses rather than options on the base class. For `_ellipsoid`', I ended up agreeing it was not necessary, since having it on the baseclass gives a little overhead only at class creation time, and in the end it is really the same thing as far as the actual calculations are concerned. But here the calculations for conversion to/from cartesian are effectively separated, making the code more complicated and giving overhead for every calculation. Generally, I think each different interpretation of the coordinates should have its own (base) class -- just like `PhysicsSpherical` is a different class from `Spherical`. I would say the rule should be that the moment `from_cartesian` and `to_cartesian` depend on any class attribute, it would be better to have a different class.
So, I think we definitely should have a separate base class for the planetocentric latitudes (`BasePlanetoCentric`?). I think we also will be better off with a separate class for West longitudes, though given the rarity of its use, perhaps this is best implemented via a mixin class that overrides `from_cartesian` and `to_cartesian` changing the sign of the longitude appropriately.
Only for `_wrap_angle`, I think it is fine to leave that on the base class -- the calculations do not depend on it, one can see it as similar to changing the class of the attributes.

mhvk commented on astropy/coordinates/representation/geodetic.py:
```diff
@@ -10,6 +11,37 @@
 
+class WestLongitudeMixin:
+    def to_cartesian(self):
+        self.lon = self.lon.value
```
I'm confused - you definitely should not change an internal attribute! How about just
```
cartesian = super().to_cartesian()
np.negative(cartesian._y, out=cartesian._y)
return cartesian
```

cmarmo replied:
I was stuck into transforming angular coordinates... now this fixes the test and avoids a double conversion... Thanks!

mhvk commented on astropy/coordinates/representation/geodetic.py:
```diff
@@ -10,6 +11,37 @@
+    @classmethod
+    def from_cartesian(cls, cart):
+        angular = super().from_cartesian(cart)
+        return cls(
```
Here, I'd again suggest to replace in-place rather than create a new instance, i.e.,
```
# Change sign of longitude in-place, possibly re-wrapping.
lon = angular._lon
np.negative(lon, out=lon)
if lon.wrap_angle != 180*u.deg:
    lon._wrap_at(lon.wrap_angle)
return angular
```

mhvk commented on astropy/coordinates/representation/geodetic.py:
```diff
@@ -10,6 +11,37 @@
+    def lon(self):
+        return self._lon
+
+    @lon.setter
```
This seems very risky; people should put in West longitudes! With my suggestions above, this is not necessary.

mhvk commented on astropy/coordinates/representation/geodetic.py:
```diff
@@ -62,12 +98,21 @@ def __init_subclass__(cls, **kwargs):
             raise AttributeError(
                 f"{cls.__name__} requires '_ellipsoid' or '_equatorial_radius' and '_flattening'."
             )
+        if not hasattr(
```
Small detail, but could do
```
if not u.Quantity(cls._wrap_angle).unit.is_equivalent(u.deg):
```
(but see my larger comment about `_wrap_angle`)

mhvk commented on astropy/coordinates/representation/geodetic.py:
```diff
@@ -62,12 +98,21 @@ def __init_subclass__(cls, **kwargs):
+        if not hasattr(
+            cls._wrap_angle, "unit"
+        ) or not cls._wrap_angle.unit.is_equivalent(u.deg):
+            raise u.UnitTypeError("Attribute _wrap_angle requires deg units.")
```
I'd write "angular units", since it is OK to give the wrap angle in radian.

mhvk commented on astropy/coordinates/representation/geodetic.py:
```diff
@@ -62,12 +98,21 @@ def __init_subclass__(cls, **kwargs):
         if height is None and not isinstance(lon, self.__class__):
             height = 0 << u.m
 
+        if not lon.unit.is_equivalent(u.deg) or not lat.unit.is_equivalent(u.deg):
```
I think this is unnecessary

mhvk commented on astropy/coordinates/representation/geodetic.py:
```diff
@@ -62,12 +98,21 @@ def __init_subclass__(cls, **kwargs):
+            raise u.UnitTypeError(
+                f"{self.__class__.__name__} requires lat and lon with units of angle."
+            )
+        lon = Longitude(lon << u.deg, wrap_angle=self._wrap_angle)
```
Why convert the input to degrees? 

Note that I'm not sure this is necessary (see larger comment). But if we stick with using `_wrap_angle`, we need to think a bit about how we can potentially make it faster - right now, the wrapping is done twice (I think).

mhvk commented on astropy/coordinates/representation/geodetic.py:
```diff
@@ -94,10 +139,114 @@ def from_cartesian(cls, cart):
             cls._equatorial_radius, cls._flattening, cart.get_xyz(xyz_axis=-1)
         )
-        return cls(lon, lat, height, copy=False)
+        return cls(
```
I don't think we should make this sense. We don't have to enforce units of degrees; if people want that, they can convert themselves.

cmarmo commented on astropy/coordinates/tests/test_representation.py:
```diff
@@ -1870,8 +1870,9 @@ def test_represent_as(self):
                 continue
-            elif name.endswith("geodetic"):
-                # TODO: Geodetic representations do not have differentials yet
+            elif True:
```
Sorry this was a mistake in a rebase.

cmarmo commented on astropy/coordinates/representation/geodetic.py:
```diff
@@ -94,10 +139,114 @@ def from_cartesian(cls, cart):
+        p_spheroid = cls._equatorial_radius * np.cos(lat)
+        z_spheroid = (cls._equatorial_radius * (1 - cls._flattening)) * np.sin(lat)
+        r_spheroid = np.sqrt(p_spheroid * p_spheroid + z_spheroid * z_spheroid)
+        height = np.where(
```
Stuck again in some stupid mental loop.. sorry, fixed.

cmarmo commented:
Thanks for your review @mhvk !
I have addressed the "obvious" fixes.
I have to better understand units and wrap angles for the rest of the comments.... I will be back ASAP.
Thanks again for your precious comments.

cmarmo commented:
The tests are failing because the class `WGS84BodycentricRepresentation`, which I have defined in order to test the ability of using `_ellipsoid` with a `BaseBodycentricRepresentation`, persists and it is used by `earth.py`.
My understanding was that the `teardown_function` would take care of "preserving the original REPRESENTATION_CLASSES", so this failure should not happen.
Anyway it also seems like the selection of a representation only based on the ellipsoid in `EarthLocation` may lead to mistakes now that `BaseBodycentricRepresentation` also uses an ellipsoid.

On the long term, I see the importance of changing `EarthLocation` to a representation, but how should I proceed on the short term? Should I remove the possibility of defining a bodycentric representation using an ellipsoid?

Any suggestion? Thanks a lot!

mhvk commented:
Simplest right now would seem to not allow `ellipsoid` for `Bodycentric`. As is, `ellipsoid` is only defined for Earth, and it is only used in `EarthLocation`, so I think this is fine.

cmarmo commented:
Hi @mhvk . I believe I have removed all the unnecessary conversions to degrees, following your comments.
I have updated the title of the pull request and the changelog to reflect what is done in the code: ie the introduction of a new ``BaseBodycentricRepresentation``.
However, I have kept the attribute `_wrap_angle` and defaulted it at 360 for `BaseBodycentricRepresentation` as this is the default recommended by Cartographic Coordinates & Rotational Elements Working Group (see https://rdcu.be/b32WL page 26)....
I'm not completely convinced in having it removed ... 

> But if we stick with using _wrap_angle, we need to think a bit about how we can potentially make it faster - right now, the wrapping is done twice (I think).


I guess this is also because I don't understand this previous comment

mhvk commented:
@cmarmo - thanks for the changes. I'm still not sold on the `_wrap_angle` attribute - essentially you are making how data is displayed part of the class, while I feel what the class should cares about is only how data is used. And for the use of a longitude, it really doesn't matter what the wrap angle is (just like its units don't matter). I still think the way forward is to define a new `Longitude180` class and use that where relevant.

However, I also think this discussion is actually distracting from the main point of this PR, which is to introduce a `BaseBodycentricRepresentation` class. How about splitting this up in pieces? I.e., here remove all stuff with `wrap_angle` and just get the basic class working.

Indeed, while I think the `WestLongitude` mixin class is the right approach, I'd even tend to put that in a separate PR as well, where it could include an update to the docs that make clearer what the use case is.

cmarmo commented:
> However, I also think this discussion is actually distracting from the main point of this PR, which is to introduce a `BaseBodycentricRepresentation` class. How about splitting this up in pieces? I.e., here remove all stuff with `wrap_angle` and just get the basic class working.
> 
> Indeed, while I think the `WestLongitude` mixin class is the right approach, I'd even tend to put that in a separate PR as well, where it could include an update to the docs that make clearer what the use case is.

I guess I'm loosing focus indeed... I need the Bodycentric representation for the frame definition, while west/east and 180/360 are not needed.... and I'm running out of time... ⏲️
Let's rescope the PR.

mhvk commented:
OK, thanks, ping me when I should look again. And apologies for not realizing this earlier!

cmarmo commented:
> And apologies for not realizing this earlier!

No problem! I'm treasuring all the comments for the follow-up! :)

cmarmo commented:
@mhvk thanks for your patience!
Is the PR acceptable in its current status?
Should I add something in the what's new section?

mhvk commented on astropy/coordinates/representation/geodetic.py:
```diff
@@ -94,10 +96,93 @@ def from_cartesian(cls, cart):
             cls._equatorial_radius, cls._flattening, cart.get_xyz(xyz_axis=-1)
         )
-        return cls(lon, lat, height, copy=False)
+        return cls(
```
Nitpick, but could you get this back to the default format? Be sure to delete the final `,` after `copy=False`, otherwise `black` will break it over several lines again.

mhvk commented on astropy/coordinates/representation/geodetic.py:
```diff
@@ -45,6 +46,7 @@ class BaseGeodeticRepresentation(BaseRepresentation):
     to quantities holding correct values (with units of length and dimensionless,
     respectively), or alternatively an ``_ellipsoid`` attribute to the relevant ERFA
     index (as passed in to `erfa.eform`).
+    Longitudes are east positive and span from 0 to 360 degrees by default.
```
Nitpick: might as well leave this out for now.

Instead, I think here is the place to mention how lon, lat, and height are defined relative to the ellipsoid.

Alternatively, you could clarify the definition in `geodetic_base_doc` above (and then *not* use it for `bodycentric`)

mhvk commented on astropy/coordinates/representation/geodetic.py:
```diff
@@ -94,10 +96,93 @@ def from_cartesian(cls, cart):
+    Subclasses need to set attributes ``_equatorial_radius`` and ``_flattening``
+    to quantities holding correct values (with units of length and dimensionless,
+    respectively).
+    Longitudes are east positive and span from 0 to 360 degrees by default.
```
Again, I'd remove this for now and instead describe how the parameters are used (and how it is different from geodetic).

Alternatively, do not use `geodetic_base_doc`, but instead write a complete docstring here that also describes how `lat` is defined.

mhvk commented on astropy/coordinates/representation/geodetic.py:
```diff
@@ -94,10 +96,93 @@ def from_cartesian(cls, cart):
+            lon,
+            lat,
+            height,
+            copy=False,
```
Again, could you bring this on a single line?

mhvk commented on docs/changes/coordinates/14851.feature.rst:
```diff
@@ -0,0 +1,2 @@
+Add ``BaseBodycentricRepresentation``, a new spheroidal representation for bodycentric
+latitudes and longitudes spanning from 0 to 360 degrees.
```
Let's remove the "spanning from 0 to 360 degrees" here too.

mhvk commented on docs/coordinates/representations.rst:
```diff
@@ -48,6 +50,13 @@ between the vertical to the surface at a specific point of the spheroid and its
+`~astropy.coordinates.BaseBodycentricRepresentation` is the coordinate representation
+recommended by the Cartographic Coordinates & Rotational Elements Working Group
+(see for example its `2019 report <https://rdcu.be/b32WL>`_): the bodicentric latitude
+and longitude are spherical latitude and longitude originated at the barycenter of the
```
originated at -> relative to

mhvk commented on docs/coordinates/representations.rst:
```diff
@@ -48,6 +50,13 @@ between the vertical to the surface at a specific point of the spheroid and its
 projection onto the equatorial plane.
 The latitude is a value ranging from -90 to 90 degrees, the longitude from 0 to 360
 degrees.
+`~astropy.coordinates.BaseBodycentricRepresentation` is the coordinate representation
```
I'd start a new paragraph here (i.e., add an empty line)

mhvk commented on docs/coordinates/representations.rst:
```diff
@@ -48,6 +50,13 @@ between the vertical to the surface at a specific point of the spheroid and its
 degrees.
+`~astropy.coordinates.BaseBodycentricRepresentation` is the coordinate representation
+recommended by the Cartographic Coordinates & Rotational Elements Working Group
+(see for example its `2019 report <https://rdcu.be/b32WL>`_): the bodicentric latitude
```
bodicentric -> bodycentric

mhvk commented on docs/coordinates/representations.rst:
```diff
@@ -48,6 +50,13 @@ between the vertical to the surface at a specific point of the spheroid and its
+recommended by the Cartographic Coordinates & Rotational Elements Working Group
+(see for example its `2019 report <https://rdcu.be/b32WL>`_): the bodicentric latitude
+and longitude are spherical latitude and longitude originated at the barycenter of the
+body, the height is the distance from the spheroid surface on the line of sight.
```
Would it be worth adding at the end something like "(i.e., the longitude and height are the same as for the geodetic representations, but the latitude is different)."

cmarmo replied:
The height is not really the same: in geodetic coordinates it is measured on a perpendicular to the surface and may have a physical meaning of elevation above the surface, while in bodycentric is more a geometrical quantity (as it is not perpendicular to the surface).

cmarmo replied:
I have clarified in the description of the geodetic coordinates.

mhvk reviewed (commented):
This looks great! There are only nitpicks and suggestions for the docstrings left. But also: do indeed add to the whatsnew entry you already created for other geodetic representations.

eerovaher reviewed (commented):
I'm letting @mhvk do the main review, but I do have a few comments.

mhvk commented on astropy/coordinates/representation/geodetic.py:
```diff
@@ -94,12 +97,85 @@ def from_cartesian(cls, cart):
+
+    Subclasses need to set attributes ``_equatorial_radius`` and ``_flattening``
+    to quantities holding correct values (with units of length and dimensionless,
+    respectively). the bodicentric latitude and longitude are spherical latitude
```
"the bodicentric" -> "The bodycentric"

mhvk commented on astropy/coordinates/representation/geodetic.py:
```diff
@@ -94,12 +97,85 @@ def from_cartesian(cls, cart):
+        d = np.hypot(p, cart.z)
+        lat = np.arctan2(cart.z, p)
+        p_spheroid = cls._equatorial_radius * np.cos(lat)
+        z_spheroid = (cls._equatorial_radius * (1 - cls._flattening)) * np.sin(lat)
```
Total nitpick, but outer parentheses are not necessary.

mhvk commented on docs/coordinates/representations.rst:
```diff
@@ -32,21 +32,32 @@ The built-in representation classes are:
-This is used internally for the standard Earth ellipsoids used in
+`~astropy.coordinates.BaseGeodeticRepresentation` is used internally for the standard
+Earth ellipsoids used in
 `~astropy.coordinates.EarthLocation`
```
Maybe combine with last line while you are making the last changes.

cmarmo replied:
Sorry I'm not sure to understand this comment....

mhvk replied:
Meant to combine the two lines to give```
Earth ellipsoids used in `~astropy.coordinates.EarthLocation`
```

mhvk commented on docs/coordinates/representations.rst:
```diff
@@ -32,21 +32,32 @@ The built-in representation classes are:
+recommended by the Cartographic Coordinates & Rotational Elements Working Group
+(see for example its `2019 report <https://rdcu.be/b32WL>`_): the bodycentric latitude
+and longitude are spherical latitude and longitude relative to the barycenter of the
+body, the height is the distance from the spheroid surface on the line of sight.
```
"the height is the distance from the spheroid surface on the line of sight" -> "height is the distance from the surface (measured radially)."

mhvk commented:
@cmarmo - separately from this PR, my sense would be to try to follow this up with adding some of the I/O and WCS system things you have been mentioning rather than worry about west longitude and wrap angles!

mhvk commented on docs/coordinates/representations.rst:
```diff
@@ -32,23 +32,39 @@ The built-in representation classes are:
+Astropy also offers a `~astropy.coordinates.BaseGeodeticRepresentation` and
+a `~astropy.coordinates.BaseBodycentricRepresentation` useful to
 create specific representations on spheroidal bodies.
-This is used internally for the standard Earth ellipsoids used in
```
I didn't mean to have this part about the internal use to be removed!

mhvk commented on astropy/coordinates/representation/geodetic.py:
```diff
@@ -110,7 +110,7 @@ class BaseBodycentricRepresentation(BaseRepresentation):
     Subclasses need to set attributes ``_equatorial_radius`` and ``_flattening``
     to quantities holding correct values (with units of length and dimensionless,
-    respectively). the bodicentric latitude and longitude are spherical latitude
+    respectively). the bodycentric latitude and longitude are spherical latitude
```
Still need to capitalize "The"

mhvk commented on docs/coordinates/representations.rst:
```diff
@@ -35,31 +35,36 @@ The built-in representation classes are:
 The latitude is a value ranging from -90 to 90 degrees, the longitude from 0 to 360
 degrees.
 
+`~astropy.coordinates.BaseGeodeticRepresentation` is used internally for the standard
```
Oops, sorry, you moved it here! This is better!

mhvk reviewed (approved):
This all looks great now. I think I'll just merge, the one uncapitalized "t" can be done at some other time!

mhvk commented:
Thanks, @cmarmo!!
